File: android/src/main/java/expo/modules/libvlcplayer/VlcPlayerModule.kt
package expo.modules.libvlcplayer

import expo.modules.libvlcplayer.enums.AudioMixingMode

import com.facebook.react.bridge.ReadableMap

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

private const val BUFFERING_EVENT = "onBuffering"
private const val PLAYING_EVENT = "onPlaying"
private const val PAUSED_EVENT = "onPaused"
private const val STOPPED_EVENT = "onStopped"
private const val ENDED_EVENT = "onEnded"
private const val REPEAT_EVENT = "onRepeat"
private const val WARN_EVENT = "onWarn"
private const val ERROR_EVENT = "onError"
private const val POSITION_CHANGED_EVENT = "onPositionChanged"
private const val LOAD_EVENT = "onLoad"
private const val BACKGROUND_EVENT = "onBackground"

val playerEvents = arrayOf(
    BUFFERING_EVENT,
    PLAYING_EVENT,
    PAUSED_EVENT,
    STOPPED_EVENT,
    ENDED_EVENT,
    REPEAT_EVENT,
    WARN_EVENT,
    ERROR_EVENT,
    POSITION_CHANGED_EVENT,
    LOAD_EVENT,
    BACKGROUND_EVENT
)

class VlcPlayerModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoLibVlcPlayer")

        OnCreate {
            VlcPlayerManager.onModuleCreated(appContext)
        }

        OnDestroy {
            VlcPlayerManager.onAppDestroyed()
        }

        View(VlcPlayerView::class) {
            Events(playerEvents)

            Prop("uri") { view: VlcPlayerView, uri: String ->
                view.uri = uri
            }

            Prop("subtitle") { view: VlcPlayerView, subtitle: ReadableMap? ->
                view.setSubtitle(subtitle)
            }

            Prop("options") { view: VlcPlayerView, options: ArrayList<String>? ->
                view.options = options ?: ArrayList<String>()
            }

            Prop("volume") { view: VlcPlayerView, volume: Int? ->
                view.setVolume(volume ?: MAX_PLAYER_VOLUME)
            }

            Prop("mute") { view: VlcPlayerView, mute: Boolean? ->
                view.setMute(mute ?: false)
            }

            Prop("rate") { view: VlcPlayerView, rate: Float? ->
                view.setRate(rate ?: DEFAULT_PLAYER_RATE)
            }

            Prop("tracks") { view: VlcPlayerView, tracks: ReadableMap? ->
                view.setTracks(tracks)
            }

            Prop("repeat") { view: VlcPlayerView, repeat: Boolean? ->
                view.setRepeat(repeat ?: false)
            }

            Prop("aspectRatio") { view: VlcPlayerView, aspectRatio: String? ->
                view.setAspectRatio(aspectRatio)
            }

            Prop("audioMixingMode") { view: VlcPlayerView, audioMixingMode: AudioMixingMode ->
                view.audioMixingMode = audioMixingMode
            }

            Prop("playInBackground") { view: VlcPlayerView, playInBackground: Boolean? ->
                view.playInBackground = playInBackground ?: false
            }

            Prop("autoplay") { view: VlcPlayerView, autoplay: Boolean? ->
                view.setAutoplay(autoplay ?: true)
            }

            OnViewDidUpdateProps { view: VlcPlayerView ->
                view.createPlayer()
            }

            OnViewDestroys { view: VlcPlayerView ->
                VlcPlayerManager.onViewDestroyed(view)
                VlcPlayerManager.unregisterView(view)
            }

            AsyncFunction("play") { view: VlcPlayerView ->
                view.play()
            }

            AsyncFunction("pause") { view: VlcPlayerView ->
                view.pause()
            }

            AsyncFunction("stop") { view: VlcPlayerView ->
                view.stop()
            }

            AsyncFunction("seek") { view: VlcPlayerView, position: Float ->
                view.seek(position)
            }
        }

        OnActivityEntersForeground {
            VlcPlayerManager.onAppForegrounded()
        }

        OnActivityEntersBackground {
            VlcPlayerManager.onAppBackgrounded()
        }
    }
}

File: ios/VlcPlayerModule.swift
import AVFoundation
import ExpoModulesCore

private let bufferingEvent = "onBuffering"
private let playingEvent = "onPlaying"
private let pausedEvent = "onPaused"
private let stoppedEvent = "onStopped"
private let endedEvent = "onEnded"
private let repeatEvent = "onRepeat"
private let errorEvent = "onError"
private let positionChangedEvent = "onPositionChanged"
private let loadEvent = "onLoad"

let playerEvents = [
    bufferingEvent,
    playingEvent,
    pausedEvent,
    stoppedEvent,
    endedEvent,
    repeatEvent,
    errorEvent,
    positionChangedEvent,
    loadEvent,
]

public class VlcPlayerModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoLibVlcPlayer")

        OnDestroy {
            VlcPlayerManager.shared.onAppDestroyed()
        }

        View(VlcPlayerView.self) {
            Events(playerEvents)

            Prop("uri") { (view: VlcPlayerView, uri: String) in
                view.setUri(uri)
            }

            Prop("subtitle") { (view: VlcPlayerView, subtitle: [String: Any]?) in
                view.setSubtitle(subtitle)
            }

            Prop("options") { (view: VlcPlayerView, options: [String]?) in
                view.setOptions(options ?? [String]())
            }

            Prop("volume") { (view: VlcPlayerView, volume: Int?) in
                view.setVolume(volume ?? maxPlayerVolume)
            }

            Prop("mute") { (view: VlcPlayerView, mute: Bool?) in
                view.setMute(mute ?? false)
            }

            Prop("rate") { (view: VlcPlayerView, rate: Float?) in
                view.setRate(rate ?? defaultPlayerRate)
            }

            Prop("tracks") { (view: VlcPlayerView, tracks: [String: Any]?) in
                view.setTracks(tracks)
            }

            Prop("time") { (view: VlcPlayerView, time: Int?) in
                view.setTime(time ?? defaultPlayerStart)
            }

            Prop("repeat") { (view: VlcPlayerView, shouldRepeat: Bool?) in
                view.setRepeat(shouldRepeat ?? false)
            }

            Prop("aspectRatio") { (view: VlcPlayerView, aspectRatio: String?) in
                view.setAspectRatio(aspectRatio)
            }

            Prop("audioMixingMode") { (view: VlcPlayerView, audioMixingMode: AudioMixingMode?) in
                view.setAudioMixingMode(audioMixingMode ?? .auto)
            }

            Prop("playInBackground") { (view: VlcPlayerView, playInBackground: Bool?) in
                view.setPlayInBackground(playInBackground ?? false)
            }

            Prop("autoplay") { (view: VlcPlayerView, autoplay: Bool?) in
                view.setAutoplay(autoplay ?? true)
            }

            OnViewDidUpdateProps { (view: VlcPlayerView) in
                view.initPlayer()
            }

            AsyncFunction("play") { (view: VlcPlayerView) in
                view.play()
            }

            AsyncFunction("pause") { (view: VlcPlayerView) in
                view.pause()
            }

            AsyncFunction("stop") { (view: VlcPlayerView) in
                view.stop()
            }

            AsyncFunction("seek") { (view: VlcPlayerView, position: Float) in
                view.seek(position)
            }
        }

        OnAppEntersForeground {
            VlcPlayerManager.shared.onAppForegrounded()
        }

        OnAppEntersBackground {
            VlcPlayerManager.shared.onAppBackgrounded()
        }
    }
}
